pyccl/halos/halo_model.py

```diff
@@ -0,0 +1,523 @@
+                     normprof=False, mass_def=None):
+    """ Solves the integral:
+
+    .. math::
+        I^1_1(k,a|u) = \\int dM\\,n(M,a)\\,b(M,a)\\,
+        \\langle u(k,a|M)\\rangle,
+
+    where :math:`n(M,a)` is the halo mass function,
+    :math:`b(M,a)` is the halo bias, and
+    :math:`\\langle u(k,a|M)\\rangle` is the halo profile as a
+    function of scale, scale factor and halo mass.
+
+    Args:
+        cosmo (:class:`~pyccl.core.Cosmology`): a Cosmology object.
+        hmc (:class:`HMCalculator`): a halo model calculator.
+        k (float or array_like): comoving wavenumber in Mpc^-1.
+        a (float or array_like): scale factor.
+        massfunc (:class:`~pyccl.halos.hmfunc.MassFunc`): a mass
+            function object.
+        hbias (:class:`~pyccl.halos.hbias.HaloBias`): a halo bias
+            object.
+        prof (:class:`~pyccl.halos.profiles.HaloProfile`): halo
+            profile.
+        normprof (bool): if `True`, this integral will be
+            normalized by :math:`I^1_0(k\\rightarrow 0,a|u)`
+            (see :meth:`~HMCalculator.mean_profile`).
+        mass_def (:class:`~pyccl.halos.massdef.MassDef`): a mass
+            definition object.
+
+    Returns:
+        float or array_like: integral values evaluated at each
+        combination of `k` and `a`. The shape of the output will
+        be `(N_a, N_k)` where `N_k` and `N_a` are the sizes of
+        `k` and `a` respectively. If `k` or `a` are scalars, the
+        corresponding dimension will be squeezed out on output.
+    """
+    a_use = np.atleast_1d(a)
+    k_use = np.atleast_1d(k)
+
+    # Check inputs
+    hmc._check_massfunc(massfunc)
+    hmc._check_hbias(hbias)
+    hmc._check_prof(prof)
+
+    na = len(a_use)
+    nk = len(k_use)
+    out = np.zeros([na, nk])
+    for ia, aa in enumerate(a_use):
+        mf = hmc._hmf(massfunc, cosmo, aa, mass_def=mass_def)
+        bf = hmc._hbf(hbias, cosmo, aa, mass_def=mass_def)
+        mbf0 = (hmc.rho0 -
+                hmc.integrator(mf * bf * hmc.mass,
+                               hmc.lmass)) / hmc.m0
+        uk = prof.fourier(cosmo, k_use, hmc.mass, aa,
+                          mass_def=mass_def).T
+        if normprof:
+            mf0 = (hmc.rho0 -
+                   hmc.integrator(mf * hmc.mass,
+                                  hmc.lmass)) / hmc.m0
+            uk0 = prof.fourier(cosmo,
+                               hmc.precision['k_min'],
+                               hmc.mass, aa,
+                               mass_def=mass_def).T
+            norm = 1. / hmc._I_0_1_from_arrays(mf, mf0, uk0)
+        else:
+            norm = 1.
+        out[ia, :] = hmc._I_1_1_from_arrays(mf, bf,
+                                            mbf0, uk) * norm
+
+    if np.ndim(a) == 0:
+        out = np.squeeze(out, axis=0)
+    if np.ndim(k) == 0:
+        out = np.squeeze(out, axis=-1)
+    return out
+
+
+def halomod_power_spectrum(cosmo, hmc, k, a, massfunc, hbias, prof,
```

Comment:
Why doesn't this function use the two functions above?

Reply:
Sorry no wrong question. 🤦‍♂ 

We should add methods like `I_2_0` or whatever right?